multiversx_sdk_cli/native_auth_client.py

```diff
@@ -0,0 +1,104 @@
+            extra_request_headers: Optional[Dict[str, str]] = None
+    ) -> None:
+        self.origin = origin
+        self.api_url = api_url
+        self.expiry_seconds = expiry_seconds
+        self.block_hash_shard = block_hash_shard
+        self.gateway_url = gateway_url
+        self.extra_request_headers = extra_request_headers
+
+
+class NativeAuthClient:
+    def __init__(self, config: Optional[NativeAuthClientConfig] = None) -> None:
+        self.config = config or NativeAuthClientConfig()
+
+    def get_token(self, address: str, token: str, signature: str) -> str:
+        encoded_address = self.encode_value(address)
+        encoded_token = self.encode_value(token)
+
+        return f"{encoded_address}.{encoded_token}.{signature}"
+
+    def initialize(self, extra_info: Dict[Any, Any] = {}) -> str:
+        block_hash = self.get_current_block_hash()
+        encoded_extra_info = self.encode_value(json.dumps(extra_info))
+        encoded_origin = self.encode_value(self.config.origin)
+
+        return f"{encoded_origin}.{block_hash}.{self.config.expiry_seconds}.{encoded_extra_info}"
+
+    def get_current_block_hash(self) -> str:
+        if self.config.gateway_url:
+            return self._get_current_block_hash_using_gateway()
+        return self._get_current_block_hash_using_api()
+
+    def _get_current_block_hash_using_gateway(self) -> str:
+        round = self._get_current_round()
+        url = f"{self.config.gateway_url}/blocks/by-round/{round}"
+        response = self._execute_request(url)
+        blocks = response["data"]["blocks"]
+        block = [b for b in blocks if b["shard"] == self.config.block_hash_shard][0]
+        return block["hash"]
+
+    def _get_current_round(self) -> int:
+        if self.config.gateway_url is None:
+            raise NativeAuthClientError("Gateway URL not set")
+
+        if self.config.block_hash_shard is None:
+            raise NativeAuthClientError("Blockhash shard not set")
+
+        url = f"{self.config.gateway_url}/network/status/{self.config.block_hash_shard}"
+        response = self._execute_request(url)
+        status = response["data"]["status"]
+
+        return status["erd_current_round"]
+
+    def _get_current_block_hash_using_api(self) -> str:
+        try:
+            url = f"{self.config.api_url}/blocks/latest?ttl={self.config.expiry_seconds}&fields=hash"
+            response = self._execute_request(url)
+            if response["hash"]:
+                return response["hash"]
+        except Exception:
+            pass
+
+        return self._get_current_block_hash_using_api_fallback()
+
+    def _get_current_block_hash_using_api_fallback(self) -> str:
+        url = f"{self.config.api_url}/blocks?size=1&fields=hash"
+
+        if self.config.block_hash_shard:
+            url += f"&shard={self.config.block_hash_shard}"
+
+        response = self._execute_request(url)
+        return response[0]["hash"]
+
+    def encode_value(self, string: str) -> str:
```

Comment:
Does this help?

https://docs.python.org/3/library/base64.html#base64.urlsafe_b64encode

`=` would have to be trimmed, nonetheless.

Reply:
Might help, but indeed `=` would have to be trimmed as well. Will leave it as it is for now.